src/box/sql/func.c

```diff
@@ -522,6 +522,168 @@ func_trim_str(struct sql_context *ctx, int argc, const struct Mem *argv)
 		ctx->is_aborted = true;
 }
 
+static uint32_t *
+prefix_table_new(struct coll *coll, const char *key, const uint32_t size)
+{
+	assert(coll != NULL);
+	uint32_t len = 0;
+	uint32_t offset = 0;
+	while (offset < size) {
+		UChar32 c;
+		U8_NEXT((uint8_t *)key, offset, size, c);
+		len++;
+	}
+	struct region *region = &fiber()->gc;
+	size_t svp = region_used(region);
+
+	uint32_t *offsets = region_alloc(region, 2 * len * sizeof(*offsets));
+	if (offsets == NULL) {
+		diag_set(OutOfMemory, 2 * len * sizeof(*offsets),
+			 "region_alloc", "offsets");
+		return NULL;
+	}
+	uint32_t *sizes = offsets + len;
+	uint32_t *table = sqlDbMallocRawNN(sql_get(), len * sizeof(*table));
+	if (table == NULL) {
+		region_truncate(region, svp);
+		return NULL;
+	}
+	offset = 0;
+	for (uint32_t i = 0; i < len; ++i) {
+		UChar32 c;
+		offsets[i] = offset;
+		U8_NEXT((uint8_t *)key, offset, size, c);
+		sizes[i] = offset - offsets[i];
+	}
+
+	table[0] = 0;
+	for (uint32_t i = 1; i < len; ++i) {
+		uint32_t k = table[i - 1];
+		int cmp_res = coll->cmp(key + offsets[k], sizes[k],
+					key + offsets[i], sizes[i], coll);
+		while (k > 0 && cmp_res != 0) {
+			k = table[k - 1];
+			cmp_res = coll->cmp(key + offsets[k], sizes[k],
+					    key + offsets[i], sizes[i], coll);
+		}
+		if (cmp_res == 0)
+			++k;
+		table[i] = k;
+	}
+	region_truncate(region, svp);
+	return table;
+}
+
+static uint32_t
+kmp_internal(struct coll *coll, const char *str, const uint32_t str_size,
+	     const char *key, const uint32_t key_size,
+	     const uint32_t *prefix_table, uint32_t *offset_out)
```

Comment:
Please rename `offset_out` to `offset` and add a comment to the function saying that this is an `out` argument. See, for example, `step_utf8_char()`. Also, add an assertion to check that `offset_out` is not NULL. What to do with this value is a problem for functions using `kmp_internal()`. And, how about using `offset` as an offset from the beginning of the line? In this case, this allows us to simplify `kmp_all()`.